// tslint-consistent-codestyle/rules/noVarBeforeReturnRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var ts = require("typescript");
var Lint = require("tslint");
var tsutils_1 = require("tsutils");
var utils_1 = require("../src/utils");
var OPTION_ALLOW_DESTRUCTURING = 'allow-destructuring';
var Rule = (function (_super) {
    tslib_1.__extends(Rule, _super);
    function Rule() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Rule.prototype.apply = function (sourceFile) {
        return this.applyWithFunction(sourceFile, walk, {
            allowDestructuring: this.ruleArguments.indexOf(OPTION_ALLOW_DESTRUCTURING) !== -1,
        });
    };
    return Rule;
}(Lint.Rules.AbstractRule));
exports.Rule = Rule;
function walk(ctx) {
    var variables;
    return ts.forEachChild(ctx.sourceFile, cbNode, cbNodeArray);
    function isUnused(node) {
        if (variables === undefined)
            variables = tsutils_1.collectVariableUsage(ctx.sourceFile);
        return variables.get(node).uses.length === 1;
    }
    function cbNode(node) {
        return ts.forEachChild(node, cbNode, cbNodeArray);
    }
    function cbNodeArray(nodes) {
        if (nodes.length === 0)
            return;
        ts.forEachChild(nodes[0], cbNode, cbNodeArray);
        for (var i = 1; i < nodes.length; ++i) {
            var node = nodes[i];
            if (tsutils_1.isReturnStatement(node)) {
                if (node.expression === undefined)
                    continue;
                if (!tsutils_1.isIdentifier(node.expression)) {
                    ts.forEachChild(node.expression, cbNode, cbNodeArray);
                    continue;
                }
                var previous = nodes[i - 1];
                if (tsutils_1.isVariableStatement(previous) && declaresVariable(previous, node.expression.text, isUnused, ctx.options))
                    ctx.addFailureAtNode(node.expression, "don't declare variable " + node.expression.text + " to return it immediately");
            }
            else {
                ts.forEachChild(node, cbNode, cbNodeArray);
            }
        }
    }
}
function declaresVariable(statement, name, isUnused, options) {
    var declarations = statement.declarationList.declarations;
    var lastDeclaration = declarations[declarations.length - 1].name;
    if (lastDeclaration.kind === ts.SyntaxKind.Identifier)
        return lastDeclaration.text === name && isUnused(lastDeclaration);
    return !options.allowDestructuring && isSimpleDestructuringForName(lastDeclaration, name, isUnused);
}
function isSimpleDestructuringForName(pattern, name, isUnused) {
    var identifiersSeen = new Set();
    var inArray = 0;
    var dependsOnVar = 0;
    return recur(pattern) === true;
    function recur(p) {
        if (p.kind === ts.SyntaxKind.ArrayBindingPattern) {
            ++inArray;
            for (var _i = 0, _a = p.elements; _i < _a.length; _i++) {
                var element = _a[_i];
                if (element.kind !== ts.SyntaxKind.OmittedExpression) {
                    var result = handleBindingElement(element);
                    if (result !== undefined)
                        return result;
                }
            }
            --inArray;
        }
        else {
            for (var _b = 0, _c = p.elements; _b < _c.length; _b++) {
                var element = _c[_b];
                var result = handleBindingElement(element);
                if (result !== undefined)
                    return result;
            }
        }
    }
    function handleBindingElement(element) {
        if (element.name.kind !== ts.SyntaxKind.Identifier) {
            if (dependsOnPrevious(element)) {
                ++dependsOnVar;
                var result = recur(element.name);
                --dependsOnVar;
                return result;
            }
            return recur(element.name);
        }
        if (element.name.text !== name)
            return void identifiersSeen.add(element.name.text);
        if (dependsOnVar !== 0)
            return false;
        if (element.dotDotDotToken) {
            if (element.parent.elements.length > 1 ||
                inArray > (element.parent.kind === ts.SyntaxKind.ArrayBindingPattern ? 1 : 0))
                return false;
        }
        else if (inArray !== 0) {
            return false;
        }
        if (element.initializer !== undefined && !utils_1.isUndefined(element.initializer))
            return false;
        return !dependsOnPrevious(element) && isUnused(element.name);
    }
    function dependsOnPrevious(element) {
        if (element.propertyName === undefined || element.propertyName.kind !== ts.SyntaxKind.ComputedPropertyName)
            return false;
        if (tsutils_1.isIdentifier(element.propertyName.expression))
            return identifiersSeen.has(element.propertyName.expression.text);
        if (tsutils_1.isLiteralExpression(element.propertyName.expression))
            return false;
        return true;
    }
}
